misc/bazel/pkg.bzl

```diff
@@ -133,17 +134,63 @@ _extract_pkg_filegroup = rule(
     } | _PLAT_DETECTION_ATTRS,
 )
 
-def _imported_zips_manifest_impl(ctx):
+_ZipInfo = provider(fields = {"zips_to_prefixes": "mapping of zip files to prefixes"})
+
+def _zip_info_impl(ctx):
+    zips = {}
+    for zip_target, prefix in ctx.attr.srcs.items():
+        for zip in zip_target.files.to_list():
+            zips[zip] = prefix
+    return [
+        _ZipInfo(zips_to_prefixes = zips),
+    ]
+
+_zip_info = rule(
+    implementation = _zip_info_impl,
+    doc = """
+        This internal rule simply instantiates a _ZipInfo provider out of `zips`.
+    """,
+    attrs = {
+        "srcs": attr.label_keyed_string_dict(
+            doc = "mapping from zip files to install prefixes",
+            allow_files = [".zip"],
+        ),
+    },
+)
+
+def _zip_info_filter_impl(ctx):
     platform = _detect_platform(ctx)
+    filtered_zips = {}
+    for zip_info in ctx.attr.srcs:
+        for zip, prefix in zip_info[_ZipInfo].zips_to_prefixes.items():
+            zip_kind, expanded_prefix = _expand_path(prefix, platform)
+            if zip_kind == ctx.attr.kind:
+                filtered_zips[zip] = expanded_prefix
+    return [
+        _ZipInfo(zips_to_prefixes = filtered_zips),
+    ]
+
+_zip_info_filter = rule(
+    implementation = _zip_info_filter_impl,
+    doc = """
+        This internal rule transforms a _ZipInfo provider so that:
+        * only zips matching `kind` are included
+        * a kind of a zip is given by its prefix: if it contains {CODEQL_PLATFORM} it is arch, otherwise it's generic
+        * in the former case, {CODEQL_PLATFORM} is expanded
+    """,
+    attrs = {
+        "srcs": attr.label_list(doc = "_ZipInfos to transform", providers = [_ZipInfo]),
+        "kind": attr.string(doc = "Which zip kind to consider", values = ["generic", "arch"]),
+    } | _PLAT_DETECTION_ATTRS,
+)
 
+def _imported_zips_manifest_impl(ctx):
     manifest = []
     files = []
-    for zip, prefix in ctx.attr.zips.items():
-        # we don't care about the kind here, as we're taking all zips together
-        _, expanded_prefix = _expand_path(prefix, platform)
-        zip_files = zip.files.to_list()
-        manifest += ["%s:%s" % (expanded_prefix, f.short_path) for f in zip_files]
-        files += zip_files
+    for zip_info in ctx.attr.srcs:
+        zip_info = zip_info[_ZipInfo]
+        manifest += ["%s:%s" % (p, z.short_path) for z, p in zip_info.zips_to_prefixes.items()]
+        files += list(zip_info.zips_to_prefixes)
```

Comment:
I had to look up the definition of `list` on a dict, it acts as `keys` function

Reply:
the python-y way to write this would be `l.extend(d)`, but I think that doesn't work in bazel (`extend` expects a list, not a generic iterable like in python). `l += d.keys()` works in bazek, but wouldn't work in python (`+=` expects a list, but in python `dict.keys` is an iterable, not a list), so that could look surprising to a python dev. `list(iterable)` to be sure to get a list is pretty idiomatic thing to do in python, and works in bazel as well so I went with it.

Reply:
I should've added: (no action required)

Reply:
I stand corrected, `extend` works with a `dict`, I'll use that then :slightly_smiling_face: